src/client/dfuse/pil4dfs/hook.c

```diff
@@ -172,77 +204,125 @@ read_map_file(char **buf, const char str_target[], char **pos)
+		D_FATAL("path_ld is too long! %d (%s)\n", ENAMETOOLONG, strerror(ENAMETOOLONG));
+		goto err;
+	}
+	D_STRNDUP(path_ld, start, end - start + 1);
 	if (path_ld == NULL) {
-		printf("Failed to allocate memory for path_ld.\nQuit\n");
-		exit(1);
+		D_FATAL("Failed to allocate memory for path_ld. %d (%s)\n", ENOMEM,
+			strerror(ENOMEM));
+		goto err;
 	}
-	memcpy(path_ld, pStart, pEnd - pStart);
-	path_ld[pEnd - pStart] = 0;
-
-	memcpy(lib_ver_str, pPos + 4, pEnd - 3 - (pPos + 4));
-	lib_ver_str[pEnd - 3 - (pPos + 4)] = 0;
-	memcpy(lib_dir_str, pStart, pPos - pStart);
-	lib_dir_str[pPos - pStart] = 0;
-
-	free(read_buff_map);
-
-	rc = asprintf(&path_libc, "%s/libc-%s.so", lib_dir_str, lib_ver_str);
-	if (rc < 0) {
-		printf("Failed to allocate memory for path_libc!\nQuit\n");
-		exit(1);
+	path_ld[end - start] = 0;
+
+	pos = strstr(read_buff_map, "libc.so");
+	if (pos == NULL)
+		/* try a different format */
+		pos = strstr(read_buff_map, "libc-2.");
+	if (pos == NULL) {
+		D_FATAL("Failed to find the path of libc.so!\n");
+		goto err;
 	}
-	if (rc >= PATH_MAX) {
-		printf("The length of path_libc is too long!\nQuit\n");
-		exit(1);
+	get_path_pos(pos, &start, &end, path_offset, read_buff_map, read_buff_map + read_size);
+	if (start == NULL || end == NULL) {
+		D_FATAL("get_path_pos() failed to determine the path for libc.so.!\n");
+		goto err;
+	}
+	if ((end - start + 1) >= PATH_MAX) {
+		D_FATAL("path_libc is too long! %d (%s)\n", ENAMETOOLONG, strerror(ENAMETOOLONG));
+		goto err;
+	}
+	/* extract the directory where libc.so is located in. */
+	D_STRNDUP(lib_dir_str, start, pos - start);
+	if (lib_dir_str == NULL) {
+		D_FATAL("Failed to allocate memory for lib_dir_str. %d (%s)\n", ENOMEM,
+			strerror(ENOMEM));
+		goto err;
+	}
+	lib_dir_str[pos - start - 1] = 0;
+	D_STRNDUP(path_libc, start, end - start + 1);
+	if (path_libc == NULL) {
+		D_FATAL("Failed to allocate memory for path_libc. %d (%s)\n", ENOMEM,
+			strerror(ENOMEM));
+		goto err;
+	}
+	path_libc[end - start] = 0;
+	D_FREE(read_buff_map);
+
+	pos = strstr(path_libc, "libc-2.");
+	if (pos) {
+		/* containing version in name. example, 2.17 */
+		memcpy(lib_ver_str, pos + 5, 4);
+		lib_ver_str[4] = 0;
 	}
 
-	rc = asprintf(&path_libpthread, "%s/libpthread-%s.so", lib_dir_str, lib_ver_str);
+	if (lib_ver_str[0]) {
+		/* with version in name */
+		rc = asprintf(&path_libpthread, "%s/libpthread-%s.so", lib_dir_str, lib_ver_str);
+	} else {
+		rc = asprintf(&path_libpthread, "%s/libpthread.so.0", lib_dir_str);
+	}
 	if (rc < 0) {
-		printf("Failed to allocate memory for path_libpthread!\nQuit\n");
-		exit(1);
+		D_FATAL("Failed to allocate memory for path_libpthread. %d (%s)\n", ENOMEM,
+			strerror(ENOMEM));
+		goto err_1;
 	}
 	if (rc >= PATH_MAX) {
-		printf("The length of path_libpthread is too long!\nQuit\n");
-		exit(1);
-	}
+		free(path_libpthread);
+		D_FATAL("path_libpthread is too long! %d (%s)\n", ENAMETOOLONG,
+			strerror(ENAMETOOLONG));
```

Comment:
As an example of what I mean above, obviously more changes than this one line would be needed.
```suggestion
		DS_ERROR(ENAMETOOLNG, "path_libpthread is too long!");
```